cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/constants.hpp>
+#include <cuspatial/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/transform.h>
+
+#include <iterator>
+#include <type_traits>
+
+namespace cuspatial {
+
+namespace detail {
+
+constexpr double EARTH_CIRCUMFERENCE_KM_PER_DEGREE = EARTH_CIRCUMFERENCE_EQUATOR_KM / 360.0;
+
+template <typename T>
+__device__ inline T midpoint(T a, T b)
+{
+  return (a + b) / 2;
+}
+
+template <typename T>
+__device__ inline T lon_to_x(T lon, T lat)
+{
+  return lon * EARTH_CIRCUMFERENCE_KM_PER_DEGREE * cos(lat * DEGREE_TO_RADIAN);
+};
+
+template <typename T>
+__device__ inline T lat_to_y(T lat)
+{
+  return lat * EARTH_CIRCUMFERENCE_KM_PER_DEGREE;
+};
+
+template <typename Location, typename T = typename Location::value_type>
+struct to_cartesian_functor {
+  to_cartesian_functor(Location origin) : _origin(origin) {}
+
+  cartesian_2d<T> __device__ operator()(Location loc)
+  {
+    return cartesian_2d<T>{lon_to_x(_origin.x - loc.x, midpoint(loc.y, _origin.y)),
+                           lat_to_y(_origin.y - loc.y)};
+  }
+
+ private:
+  Location _origin{};
+};
+
+}  // namespace detail
+
+template <class InputIt, class OutputIt, class Location>
+OutputIt lonlat_to_cartesian(InputIt lon_lat_first,
+                             InputIt lon_lat_last,
+                             OutputIt xy_first,
+                             Location origin,
```

Comment:
If `Location` can only be `lonlat_2d<T>`, why not just make it `lonlat_2d<T>`?

Reply:
Good point. Fixed.